fendermint/vm/snapshot/src/state.rs

```diff
@@ -70,6 +74,59 @@ impl SnapshotItem {
 
         Ok(content)
     }
+
+    /// Import a snapshot into the blockstore.
+    pub async fn import<BS>(&self, store: BS, validate: bool) -> anyhow::Result<Snapshot<BS>>
+    where
+        BS: Blockstore + Send + 'static,
+    {
+        let parts = manifest::list_parts(self.snapshot_dir.join(PARTS_DIR_NAME))
+            .context("failed to list snapshot parts")?;
+
+        // 1. Restore the snapshots into a complete `snapshot.car` file.
+        let car_path = self.snapshot_dir.join(SNAPSHOT_FILE_NAME);
+        let mut car_file = File::create(&car_path).context("failed to create CAR file")?;
+
+        for part in parts {
+            let mut part_file = File::open(&part).with_context(|| {
+                format!("failed to open snapshot part {}", part.to_string_lossy())
+            })?;
+
+            io::copy(&mut part_file, &mut car_file)?;
+        }
+
+        // 2. Import the contents.
+        let result = Snapshot::read_car(&car_path, store).await;
+
+        // 3. Remove the restored file.
+        std::fs::remove_file(&car_path).context("failed to remove CAR file")?;
+
+        let snapshot = result.context("failed to import the snapshot into the blockstore")?;
+
+        // 4. See if we actually imported what we thought we would.
```

Comment:
What would happen if the validation fail and we've already imported the snapshot? I guess you would need to import a valid snapshot, and the invalid blocks would still be stored in the blockstore, right? I don't think is critical at this point, but worth tracking it somewhere, as several import failures could make the blockstore size increase  significantly. Can't we do the validation before importing the CAR?

Reply:
That's right, the unwanted blocks would be left in the blockstore. Because they have been imported into the state store directly, the only way to remove them would be to implement garbage collection. 

I actually included such thoughts into this ticket as loading the data into a staging area inside RocksDB (a separate namespace/collection) and only if successful copying it over to their final destination, or dropping the collection if the import failed. However, I don't think I have the necessary API exposed at the moment and didn't want to complicate things while we don't even have the happy path working. 

> Can't we do the validation before importing the CAR?

To be honest I was only following what Will's `Snapshot` lets me do easily. I see it uses `load_car_unchecked` which doesn't even verify CIDs; I'll change that. By the looks of it it's generally not easy to validate a CAR file: in our case the only CID in the header is the one of the metadata record, which only points at the CID of some data records, which is a tuple that is an internal implementation detail of the `Snapshot`. It is actually the [last thing](https://github.com/consensus-shipyard/fendermint/blob/3cf934c5fe7388814396b373184625938683b45d/fendermint/vm/interpreter/src/fvm/state/snapshot.rs#L192) it puts in the file I think, but to get there you'd have to enumerate everything and then parse it correctly - and this is just the V1 snapshot. 

So I landed on the "trust, but verify" approach.

I checked what kind of validations I noted down in https://github.com/consensus-shipyard/fendermint/issues/364 but they are also post-import.